File: .skills/openclaw-skills/skills/0xnyk/xint/lib/bookmarks.ts
/**
 * Bookmarks — fetch and display user bookmarks via OAuth 2.0.
 * Requires: auth setup (PKCE flow) to have been run first.
 */

import { writeFileSync } from "fs";
import { join } from "path";
import { BASE, FIELDS, oauthGet, parseTweets, parseSince, sleep, sortBy, dedupe } from "./api";
import type { Tweet } from "./api";
import { getValidToken, loadTokens } from "./oauth";
import * as cache from "./cache";
import * as fmt from "./format";
import { trackCost } from "./costs";

const SKILL_DIR = join(import.meta.dir, "..");
const DRAFTS_DIR = join(SKILL_DIR, "data", "exports");
const BOOKMARK_CACHE_TTL_MS = 5 * 60 * 1000; // 5 minutes

/**
 * Fetch bookmarks via GET /2/users/{id}/bookmarks with pagination.
 * Max 800 bookmarks total (API limit), 100 per page.
 */
export async function fetchBookmarks(
  userId: string,
  accessToken: string,
  maxTotal: number = 100
): Promise<Tweet[]> {
  const allTweets: Tweet[] = [];
  let nextToken: string | undefined;
  const perPage = Math.min(maxTotal, 100);
  const maxPages = Math.ceil(Math.min(maxTotal, 800) / perPage);

  for (let page = 0; page < maxPages; page++) {
    const pagination = nextToken ? `&pagination_token=${nextToken}` : "";
    const url = `${BASE}/users/${userId}/bookmarks?max_results=${perPage}&${FIELDS}${pagination}`;

    const raw = await oauthGet(url, accessToken);
    const tweets = parseTweets(raw);
    allTweets.push(...tweets);

    if (allTweets.length >= maxTotal) break;

    nextToken = (raw as any).meta?.next_token;
    if (!nextToken) break;
    if (page < maxPages - 1) await sleep(350);
  }

  return allTweets.slice(0, maxTotal);
}

/**
 * CLI handler for the bookmarks command.
 */
export async function cmdBookmarks(args: string[]): Promise<void> {
  // Parse flags
  const flags = new Set<string>();
  const opts: Record<string, string> = {};

  const positional: string[] = [];
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg.startsWith("--")) {
      const name = arg.slice(2);
      // Boolean flags
      if (["json", "markdown", "save", "no-cache"].includes(name)) {
        flags.add(name);
      } else if (i + 1 < args.length) {
        opts[name] = args[++i];
      }
    } else {
      positional.push(arg);
    }
  }

  const limit = parseInt(opts.limit || "20");
  const sinceStr = opts.since;
  const queryFilter = opts.query;
  const asJson = flags.has("json");
  const asMarkdown = flags.has("markdown");
  const save = flags.has("save");
  const noCache = flags.has("no-cache");

  // Get OAuth token (auto-refresh if needed)
  const accessToken = await getValidToken();
  const tokens = loadTokens()!;

  // Check cache
  const cacheKey = `bookmarks:${tokens.user_id}`;
  const cacheParams = `limit=${limit}`;

  if (!noCache) {
    const cached = cache.get(cacheKey, cacheParams, BOOKMARK_CACHE_TTL_MS);
    if (cached) {
      console.error(`(cached — ${cached.length} bookmarks)`);
      outputBookmarks(cached, { limit, sinceStr, queryFilter, asJson, asMarkdown, save });
      return;
    }
  }

  // Fetch bookmarks — request more than limit to allow for client-side filtering
  const fetchCount = Math.min(
    sinceStr || queryFilter ? Math.max(limit * 3, 100) : limit,
    800
  );

  console.error(`Fetching bookmarks for @${tokens.username}...`);
  const tweets = await fetchBookmarks(tokens.user_id, accessToken, fetchCount);

  // Track cost
  trackCost("bookmarks", `/2/users/me/bookmarks`, tweets.length);

  if (tweets.length === 0) {
    console.log("No bookmarks found.");
    return;
  }

  // Cache raw results
  cache.set(cacheKey, cacheParams, tweets);

  outputBookmarks(tweets, { limit, sinceStr, queryFilter, asJson, asMarkdown, save });
}

function outputBookmarks(
  tweets: Tweet[],
  opts: {
    limit: number;
    sinceStr?: string;
    queryFilter?: string;
    asJson: boolean;
    asMarkdown: boolean;
    save: boolean;
  }
): void {
  let filtered = [...tweets];

  // Client-side --since filter
  if (opts.sinceStr) {
    const sinceTs = parseSince(opts.sinceStr);
    if (sinceTs) {
      const sinceMs = new Date(sinceTs).getTime();
      filtered = filtered.filter((t) => new Date(t.created_at).getTime() >= sinceMs);
    }
  }

  // Client-side --query text filter (case-insensitive substring match)
  if (opts.queryFilter) {
    const q = opts.queryFilter.toLowerCase();
    filtered = filtered.filter(
      (t) =>
        t.text.toLowerCase().includes(q) ||
        t.username.toLowerCase().includes(q) ||
        t.hashtags.some((h) => h.toLowerCase().includes(q))
    );
  }

  filtered = dedupe(filtered);
  const shown = filtered.slice(0, opts.limit);

  if (shown.length === 0) {
    console.log("No bookmarks match the filters.");
    return;
  }

  // Output
  if (opts.asJson) {
    console.log(JSON.stringify(shown, null, 2));
  } else if (opts.asMarkdown) {
    const md = fmt.formatResearchMarkdown("Bookmarks", shown, {
      queries: ["bookmarks"],
    });
    console.log(md);
  } else {
    console.log(fmt.formatResultsTelegram(shown, { query: "Bookmarks", limit: opts.limit }));
  }

  // Save
  if (opts.save) {
    const date = new Date().toISOString().split("T")[0];
    const path = join(DRAFTS_DIR, `x-bookmarks-${date}.md`);
    const md = fmt.formatResearchMarkdown("Bookmarks", shown, {
      queries: ["bookmarks"],
    });
    writeFileSync(path, md);
    console.error(`\nSaved to ${path}`);
  }

  // Stats
  const filterNote = filtered.length !== tweets.length
    ? ` (${filtered.length} after filters)`
    : "";
  console.error(`\n📑 ${tweets.length} bookmarks fetched${filterNote}, showing ${shown.length}`);
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/commands/bookmarks.rs
use anyhow::Result;
use std::fs;

use crate::api::twitter;
use crate::auth::oauth;
use crate::cli::BookmarksArgs;
use crate::client::XClient;
use crate::config::Config;
use crate::costs;
use crate::format;

pub async fn run(args: &BookmarksArgs, config: &Config, client: &XClient) -> Result<()> {
    let client_id = config.require_client_id()?;
    let (access_token, tokens) =
        oauth::get_valid_token(client, &config.tokens_path(), client_id).await?;

    let cache_key = format!("bookmarks:{}", tokens.user_id);
    let cache_params = format!("limit={}", args.limit);
    let cache_ttl = 5 * 60 * 1000u64; // 5 minutes

    let cached: Option<Vec<crate::models::Tweet>> = if args.no_cache {
        None
    } else {
        crate::cache::get(&config.cache_dir(), &cache_key, &cache_params, cache_ttl)
    };

    let all_tweets = if let Some(cached) = cached {
        eprintln!("(cached — {} bookmarks)", cached.len());
        cached
    } else {
        let fetch_count = if args.since.is_some() || args.query.is_some() {
            (args.limit * 3).clamp(100, 800)
        } else {
            args.limit.min(800)
        };

        eprintln!("Fetching bookmarks for @{}...", tokens.username);

        let mut all = Vec::new();
        let mut next_token: Option<String> = None;
        let per_page = fetch_count.min(100);
        let max_pages = fetch_count.div_ceil(per_page).min(8);

        for page in 0..max_pages {
            let pagination = match &next_token {
                Some(t) => format!("&pagination_token={t}"),
                None => String::new(),
            };
            let path = format!(
                "users/{}/bookmarks?max_results={}&{}{}",
                tokens.user_id,
                per_page,
                crate::client::FIELDS,
                pagination
            );

            let raw = client.oauth_get(&path, &access_token).await?;
            let tweets = twitter::parse_tweets(&raw);
            all.extend(tweets);

            if all.len() >= fetch_count {
                break;
            }
            next_token = raw.meta.and_then(|m| m.next_token);
            if next_token.is_none() {
                break;
            }
            if page < max_pages - 1 {
                crate::client::rate_delay().await;
            }
        }

        costs::track_cost(
            &config.costs_path(),
            "bookmarks",
            "/2/users/bookmarks",
            all.len() as u64,
        );

        crate::cache::set(&config.cache_dir(), &cache_key, &cache_params, &all);
        all
    };

    // Client-side filtering
    let mut filtered = all_tweets;

    if let Some(ref since) = args.since {
        if let Some(ts) = twitter::parse_since(since) {
            if let Ok(since_dt) = chrono::DateTime::parse_from_rfc3339(&ts) {
                let since_ms = since_dt.timestamp_millis();
                filtered.retain(|t| {
                    chrono::DateTime::parse_from_rfc3339(&t.created_at)
                        .map(|d| d.timestamp_millis() >= since_ms)
                        .unwrap_or(true)
                });
            }
        }
    }

    if let Some(ref q) = args.query {
        let q_lower = q.to_lowercase();
        filtered.retain(|t| {
            t.text.to_lowercase().contains(&q_lower)
                || t.username.to_lowercase().contains(&q_lower)
                || t.hashtags
                    .iter()
                    .any(|h| h.to_lowercase().contains(&q_lower))
        });
    }

    filtered = twitter::dedupe(filtered);
    let shown: Vec<_> = filtered.into_iter().take(args.limit).collect();

    if shown.is_empty() {
        println!("No bookmarks match the filters.");
        return Ok(());
    }

    if args.json {
        format::print_json_pretty_filtered(&shown)?;
    } else if args.markdown {
        let md = format::format_research_markdown("Bookmarks", &shown, &["bookmarks"]);
        println!("{md}");
    } else {
        println!(
            "{}",
            format::format_results_terminal(&shown, Some("Bookmarks"), args.limit)
        );
    }

    if args.save {
        let exports_dir = config.exports_dir();
        fs::create_dir_all(&exports_dir)?;
        let date = chrono::Utc::now().format("%Y-%m-%d").to_string();
        let path = exports_dir.join(format!("x-bookmarks-{date}.md"));
        let md = format::format_research_markdown("Bookmarks", &shown, &["bookmarks"]);
        fs::write(&path, &md)?;
        eprintln!("\nSaved to {}", path.display());
    }

    Ok(())
}
